File: app/src/main/java/com/ebnbin/gank/feature/days/dayviewpager/DayViewPagerFragment.java
package com.ebnbin.gank.feature.days.dayviewpager;

import android.graphics.drawable.ColorDrawable;
import android.graphics.drawable.Drawable;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.view.ViewPager;
import android.support.v7.app.ActionBar;
import android.view.View;

import com.ebnbin.eb.util.EBUtil;
import com.ebnbin.eb.util.Timestamp;
import com.ebnbin.ebapplication.context.EBActionBarFragment;
import com.ebnbin.ebapplication.context.EBFragment;
import com.ebnbin.ebapplication.net.NetModelCallback;
import com.ebnbin.gank.R;

import okhttp3.Call;
import okhttp3.Response;

/**
 * 用 {@link ViewPager} 展示多个 {@link com.ebnbin.gank.feature.days.day.DayFragment}.
 */
public final class DayViewPagerFragment extends EBFragment {
    private ViewPager mDaysViewPager;

    @Override
    protected int overrideContentViewLayout() {
        return R.layout.days_day_view_pager_fragment;
    }

    private DayViewPagerPagerAdapter mDayViewPagerPagerAdapter;

    @Override
    public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        mDaysViewPager = view.findViewById(R.id.days);

        mDayViewPagerPagerAdapter = new DayViewPagerPagerAdapter(getChildFragmentManager());
        mDaysViewPager.setAdapter(mDayViewPagerPagerAdapter);

        mDaysViewPager.setOffscreenPageLimit(1);

        int marginPixels = getResources().getDimensionPixelSize(R.dimen.days_page_margin);
        mDaysViewPager.setPageMargin(marginPixels);

        Drawable d = new ColorDrawable(EBUtil.INSTANCE.getColorAttr(getContext(), R.attr.ebColorPlaceholder));
        mDaysViewPager.setPageMarginDrawable(d);

        mDaysViewPager.addOnPageChangeListener(new ViewPager.OnPageChangeListener() {
            private boolean mDragged = false;

            @Override
            public void onPageScrolled(int position, float positionOffset, int positionOffsetPixels) {
            }

            @Override
            public void onPageSelected(int position) {
                setTitle(position);

                EBActionBarFragment actionBarFragment = getActionBarParentFragment();
                if (actionBarFragment != null && mDragged) {
                    mDragged = false;

                    actionBarFragment.getAppBarLayout().setExpanded(true, true);
                }

                mDayViewPagerPagerAdapter.getFuliUrl(position, new DayViewPagerPagerAdapter.FuliUrlCallback() {
                    @Override
                    public void onGetFuliUrl(@Nullable String fuliUrl) {
                        super.onGetFuliUrl(fuliUrl);

                        // TODO: 福利 image.
                    }
                });
            }

            @Override
            public void onPageScrollStateChanged(int state) {
                if (state == ViewPager.SCROLL_STATE_DRAGGING) {
                    mDragged = true;
                }
            }
        });
    }

    /**
     * Sets title of {@link ActionBar}.
     */
    private void setTitle(int position) {
        ActionBar actionBar = getEbActivity().getSupportActionBar();
        if (actionBar == null) {
            return;
        }

        Timestamp timestamp = mDayViewPagerPagerAdapter.timestamps.get(position);
        if (timestamp == null) {
            return;
        }

        String title = getString(R.string.days_title, timestamp.getYear(), timestamp.getMonth(), timestamp.getDay());
        actionBar.setTitle(title);
    }

    //*****************************************************************************************************************
    // Instance state.

    private static final String STATE_HISTORY_MODEL = "history_model";
    private static final String STATE_CURRENT_ITEM = "current_item";

    @Override
    public void onViewStateRestored(@Nullable Bundle savedInstanceState) {
        super.onViewStateRestored(savedInstanceState);

        if (!restoreHistory(savedInstanceState)) {
            netGetHistory();
        }
    }

    /**
     * 恢复 {@link #mHistoryModel}.
     *
     * @return 是否成功恢复.
     */
    public boolean restoreHistory(@Nullable Bundle savedInstanceState) {
        if (savedInstanceState == null) {
            return false;
        }

        mHistoryModel = (HistoryModel) savedInstanceState.getSerializable(STATE_HISTORY_MODEL);
        if (mHistoryModel == null) {
            return false;
        }

        mDayViewPagerPagerAdapter.setHistoryModel(mHistoryModel);

        int defaultCurrentItem = mDayViewPagerPagerAdapter.getCount() - 1;
        int currentItem = savedInstanceState.getInt(STATE_CURRENT_ITEM, defaultCurrentItem);
        if (currentItem >= 0 && currentItem < mDayViewPagerPagerAdapter.getCount()) {
            mDaysViewPager.setCurrentItem(currentItem, false);
        }

        return true;
    }

    @Override
    public void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);

        if (outState == null) {
            return;
        }

        outState.putSerializable(STATE_HISTORY_MODEL, mHistoryModel);
        outState.putInt(STATE_CURRENT_ITEM, mDaysViewPager.getCurrentItem());
    }

    //*****************************************************************************************************************
    // Net.

    private HistoryModel mHistoryModel;

    /**
     * Gets history and sets data to {@link DayViewPagerPagerAdapter}.
     */
    private void netGetHistory() {
        String url = "http://gank.io/api/day/history";
        netGet(url, new NetModelCallback<HistoryModel>() {
            @Override
            public void onSuccess(@NonNull Call call, @NonNull HistoryModel model, @NonNull Response response,
                    @NonNull byte[] byteArray) {
                super.onSuccess(call, model, response, byteArray);

                mHistoryModel = model;
                mDayViewPagerPagerAdapter.setHistoryModel(mHistoryModel);

                int item = mDayViewPagerPagerAdapter.getCount() - 1;
                if (item >= 0 && item < mDayViewPagerPagerAdapter.getCount()) {
                    mDaysViewPager.setCurrentItem(item, false);
                }
            }
        });
    }
}

File: app/src/main/java/com/ebnbin/gank/feature/days/dayviewpager/DayViewPagerFragment.kt
package com.ebnbin.gank.feature.days.dayviewpager

import android.graphics.drawable.ColorDrawable
import android.os.Bundle
import android.support.v4.view.ViewPager
import android.support.v7.app.ActionBar
import android.view.MenuItem
import android.view.View
import com.ebnbin.eb.util.EBUtil
import com.ebnbin.eb.util.Timestamp
import com.ebnbin.ebapplication.context.EBFragment
import com.ebnbin.ebapplication.net.NetModelCallback
import com.ebnbin.gank.R
import com.ebnbin.gank.feature.days.HistoryModel
import com.ebnbin.recyclercalendarview.recyclerdatepicker.RecyclerDatePickerSupportDialogFragment
import okhttp3.Call
import okhttp3.Response

/**
 * 用 [ViewPager] 展示多个 [com.ebnbin.gank.feature.days.day.DayFragment].
 */
class DayViewPagerFragment : EBFragment(), RecyclerDatePickerSupportDialogFragment.Callback {
    override fun overrideContentViewLayout(): Int {
        return R.layout.days_day_view_pager_fragment
    }

    private val daysViewPager: ViewPager by lazy {
        stateView.findViewById(R.id.days) as ViewPager
    }

    private val pagerAdapter: DayViewPagerPagerAdapter by lazy {
        DayViewPagerPagerAdapter(childFragmentManager)
    }

    override fun onViewCreated(view: View?, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        daysViewPager.adapter = pagerAdapter
        daysViewPager.offscreenPageLimit = 1
        daysViewPager.pageMargin = resources.getDimensionPixelSize(R.dimen.days_page_margin)
        daysViewPager.setPageMarginDrawable(ColorDrawable(EBUtil.getColorAttr(context, R.attr.ebColorPlaceholder)))
        daysViewPager.addOnPageChangeListener(object : ViewPager.OnPageChangeListener {
            private var dragged = false

            override fun onPageScrolled(position: Int, positionOffset: Float, positionOffsetPixels: Int) {}

            override fun onPageSelected(position: Int) {
                setTitle(position)

                if (actionBarParentFragment != null && dragged) {
                    dragged = false

                    actionBarParentFragment!!.appBarLayout.setExpanded(true, true)
                }

                pagerAdapter.getFirstFuli(position, object : DayViewPagerPagerAdapter.FirstFuliCallback() {
                    override fun onGetFirstFuli(url: String?) {
                        super.onGetFirstFuli(url)

                        // TODO: 福利 image.
                    }
                })
            }

            override fun onPageScrollStateChanged(state: Int) {
                if (state == ViewPager.SCROLL_STATE_DRAGGING) {
                    dragged = true
                }
            }

            /**
             * Sets title of [ActionBar].
             */
            private fun setTitle(position: Int) {
                val toolbar = actionBarParentFragment?.toolbar ?: return

                val timestamp = pagerAdapter.timestamps[position]
                val title = getString(R.string.days_title, timestamp.year, timestamp.month, timestamp.day)

                toolbar.title = title
            }
        })

        initActionBar()
    }

    override fun onViewStateRestored(savedInstanceState: Bundle?) {
        super.onViewStateRestored(savedInstanceState)

        if (!restoreHistory(savedInstanceState)) {
            netGetHistory()
        }
    }

    /**
     * 恢复 [historyModel].

     * @return 是否成功恢复.
     */
    fun restoreHistory(savedInstanceState: Bundle?): Boolean {
        if (savedInstanceState == null) {
            return false
        }

        historyModel = savedInstanceState.getSerializable(STATE_HISTORY_MODEL) as HistoryModel?
        if (historyModel == null) {
            return false
        }

        pagerAdapter.setData(historyModel!!.timestamps)

        val defaultCurrentItem = pagerAdapter.count - 1
        val currentItem = savedInstanceState.getInt(STATE_CURRENT_ITEM, defaultCurrentItem)
        if (currentItem >= 0 && currentItem < pagerAdapter.count) {
            daysViewPager.setCurrentItem(currentItem, false)
        }

        return true
    }

    override fun onSaveInstanceState(outState: Bundle?) {
        super.onSaveInstanceState(outState)

        if (outState == null) {
            return
        }

        outState.putSerializable(STATE_HISTORY_MODEL, historyModel)
        outState.putInt(STATE_CURRENT_ITEM, daysViewPager.currentItem)
    }

    //*****************************************************************************************************************
    // Net.

    private var historyModel: HistoryModel? = null

    /**
     * Gets history and sets data to [DayViewPagerPagerAdapter].
     */
    private fun netGetHistory() {
        val url = "http://gank.io/api/day/history"
        netGet(url, object : NetModelCallback<HistoryModel>() {
            override fun onSuccess(call: Call, model: HistoryModel, response: Response,
                    byteArray: ByteArray) {
                super.onSuccess(call, model, response, byteArray)

                historyModel = model
                pagerAdapter.setData(historyModel!!.timestamps)

                if (calendarMenuItem != null) {
                    calendarMenuItem!!.isVisible = true
                }

                val item = pagerAdapter.count - 1
                if (item >= 0 && item < pagerAdapter.count) {
                    daysViewPager.setCurrentItem(item, false)
                }
            }
        })
    }

    //*****************************************************************************************************************
    // ActionBar.

    private var calendarMenuItem: MenuItem? = null

    private fun initActionBar() {
        val toolbar = actionBarParentFragment?.toolbar ?: return

        toolbar.inflateMenu(R.menu.fragment_days_day_view_pager)
        calendarMenuItem = toolbar.menu.findItem(R.id.calendar)
        calendarMenuItem!!.isVisible = historyModel != null
        toolbar.menu.findItem(R.id.calendar).setOnMenuItemClickListener({
            RecyclerDatePickerSupportDialogFragment.showDialog(childFragmentManager, historyModel!!.timestamps,
                    pagerAdapter.timestamps[daysViewPager.currentItem])
            true
        })
    }

    //*****************************************************************************************************************

    override fun onSelected(date: Timestamp) {
        val position = pagerAdapter.timestamps.indexOf(date)
        if (position != -1) {
            daysViewPager.setCurrentItem(position, true)
        }
    }

    companion object {
        //*************************************************************************************************************
        // Instance state.

        private val STATE_HISTORY_MODEL = "history_model"
        private val STATE_CURRENT_ITEM = "current_item"
    }
}
